### PR TITLE
Spec `ExecutionState::STOP`

### PR DESCRIPTION
**Pull request by han0110**
This PR aims to spec out `ExecutionState::STOP`.

## TODO

- [x] Handle implicit `STOP` when `program_counter == code.length` (discussion in #102).

### CODE REVIEW

**Review by ed255 — APPROVED** [review]

File: specs/opcode/00STOP.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+The `STOP` opcode terminates the call, then:
+
+1. If it's an root call, it ends transaction.
+2. Otherwise, it restores caller's context and switch to it.
+
+### Circuit behavior
+
+The circuit first checks the `result` in call context is indeed success. Then:
+
+1. If it's an root call, it handles the gas refund and coinbase reward, then goes to next transaction by checking the next step has increased `tx_id`.
```
[the remark this answers is not in the document]
Comment (ed255) [reply specs/opcode/00STOP.md:16]: I hit approve by mistake, so I had to add a comment to change my review status.

**Review by ed255 — CHANGES_REQUESTED** [review]

File: specs/opcode/00STOP.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (han0110) [reply specs/opcode/00STOP.md]: Yes, it's handled by `EndTx`. Fixed.

**Review by ed255 — APPROVED** [review]
LGTM!

**Review by ed255 — APPROVED** [review]
I've left two small comments but everything looks correct to me :)
I'm approving the PR already, feel free to check the comments and discuss!

**Comment by ed255** [comment]
BTW, the current branch conflicts seem to come from https://github.com/appliedzkp/zkevm-specs/pull/185 and should be very easy to resolve.

File: src/zkevm_specs/evm/execution/stop.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        #   STOP returns empty bytes as deployment code, but when it's an internal creation call,
+        #   the code_hash of callee must already be random linear combination of EMPTY_CODE_HASH,
+        #   which doesn't need any update here.
+        # 2. not is_create:
+        #   STOP returns empty bytes as return_data, which doesn't affect caller's memory at all.
+        # So we only need to restore caller's state as finishing this call.
+
+        # Restore caller state to next StepState
+        instruction.step_state_transition_to_restored_context(
+            rw_counter=Transition.delta(13),
```
[the remark this answers is not in the document]
Comment (han0110) [reply src/zkevm_specs/evm/execution/stop.py:50]: Yes, this approach is much easier to follow, let me update it!

[the remark this answers is not in the document]
Comment (han0110) [reply src/zkevm_specs/evm/execution/stop.py]: Fixed in [`7546f8b`](https://github.com/appliedzkp/zkevm-specs/pull/100/commits/7546f8b3360320052db1d12204f3f5b21783fb1e)

**Review by icemelon — COMMENTED** [review]
LGTM. Just two minor comments.

**Review by icemelon — APPROVED** [review]

File: src/zkevm_specs/evm/execution/calldatacopy.py [hunk 3]
```diff
@@ -37,7 +37,6 @@ def calldatacopy(instruction: Instruction):
 
     # When length != 0, constrain the state in the next execution state CopyToMemory
     if instruction.is_zero(length) == FQ(0):
-        assert instruction.next is not None
```
[the remark this answers is not in the document]
Comment (icemelon) [reply src/zkevm_specs/evm/execution/calldatacopy.py:40]: Nope, just curious.

**Pull request merged**